#  Master Thesis - Transcriptome Analyses in Aspidoscelis neomexicanus

###### ================================================================================
###### OVERVIEW
+ 4. [ Installation of required software  ](#install)
    * 4.1 [ General software requirements ](#gensof)
    * 4.2 [ Preprocessing ](#prepro)
    * 4.3 [ Transcriptome Reconstruction ](#trans-recon)
    * 4.4 [ Transcriptome Annotation ](#trans-anno)
###### ================================================================================

Go back to [README_Master_Thesis.md](README_Master_Thesis.md)

 <a name="gensof"></a>
## 4.1 General software requirements
These tools are needed to set up the environment and run the pipeline.

<a name="miniforge3"></a>
**4.1.1 [miniforge3](https://github.com/conda-forge/miniforge)**
Used to create environments with conda and mamba for easier package managment and software installation.

<a name="next"></a>
**4.1.2 [nextflow](https://www.nextflow.io/docs/latest/install.html)**
Workflow manager to run the analysis pipeline in a reproducible way.

<a name="docker"></a>
**4.1.3 [docker](https://docs.docker.com/engine/install/ubuntu/)**
Allows to run software in containers. Required for using the provided Docker Image.

<a name="xxx"></a>
**4.1.4 Java[conda_java20_env.yml](conda_java20_env.yml)**
OpenJDK version 20. Required to run Nextflow pipeline. Provided is a yml file to install it in a conda evironment.

<a name="prepro"></a>
## 4.2 Preprocessing 
These tools are used before transcriptome reconstruction, to basecall, evaluate and refine the raw sequencing data.

<a name="dorado"></a>
**4.2.1 [dorado](https://github.com/nanoporetech/dorado)**
Used to convert raw ONT .pod5 files into .ubam basecalled reads.

<a name="pycoQC"></a>
**4.2.2 [pycoQC](https://a-slide.github.io/pycoQC/installation/)**
Generates metrics and interactive plots for ONT seqeuncing data.

<a name="NanoPlot"></a>
**4.2.3 [NanoPlot](https://github.com/wdecoster/NanoPlot)**
Generates metrics and plots for ONT seqeuncing data.

<a name="NanoComp"></a>
**4.2.4 [NanoComp](https://github.com/wdecoster/nanocomp)**
Compares multiple runs of long read sequencing data and alignments based on QC metrics and plots.

<a name="trans-recon"></a>
## 4.3 Transcriptome Reconstruction 
these tools are used to reconstruct the transcriptome from the ONT dRNA sequencing reads using a reference genome file.
<a name="minimap2"></a>
**4.3.1 [minimap2](https://lh3.github.io/minimap2/minimap2.html)**
Aligns the raw seqeuncign reads to the reference genome in spliced mode.

<a name="samtools"></a>
**4.3.2 [samtools](https://www.htslib.org/doc/samtools.html)**
Sorts and indexes the alignment files (.bam/.sam) generated by minimap2.

<a name="stringtie2"></a>
**4.3.3 [stringtie2](https://github.com/skovaka/stringtie2)**
Reconstruct the transcriptome from the minimap2 mappings and outputs a GTF file containing transcript structure information.

<a name="gffread"></a>
**4.3.4 [gffread](https://github.com/gpertea/gffread)**
Extracts transcript seqeunces from the genome assembly using the stringtie GTF output and generated a fasta file.

<a name="trans-anno"></a>
**4.4 Transcriptome Annotation**
These tools annotate the reconstructed transcripts with protein function information.

<a name="eggNOG"></a>
**4.3.1 [eggNOG](https://github.com/eggnogdb/eggnog-mapper/wiki/eggNOG-mapper-v2.1.5-to-v2.1.13)**
Annotates transcripts using orthology-based functional information.

<a name="uniprot"></a>
**4.3.3 [UniProt]() - xxx**

## 4.5 Comparative Analyses
????

