#  Software Requirements

###### ================================================================================
###### OVERVIEW
+ 4. [ Installation of required software  ](#install)
    * 4.1 [ General software requirements ](#gensof)
    * 4.2 [ Preprocessing ](#prepro)
    * 4.3 [ Transcriptome Reconstruction ](#trans-recon)
    * 4.4 [ Transcriptome Annotation ](#trans-anno)
###### ================================================================================
Go back to [README_Master_Thesis.md](README_Master_Thesis.md)


 <a name="gensof"></a>
## 4.1 General software requirements
These tools are needed to set up the environment and run the pipeline.
<a name="miniforge3"></a>
### 4.1.1 [miniforge3](https://github.com/conda-forge/miniforge)
Used to create environments with conda and mamba for easier package management and software installation.
### 4.1.2 [nextflow](https://github.com/nextflow-io/nextflow)
Workflow manager to run the analysis pipeline in a reproducible way.
<a name="docker"></a>
### 4.1.3 [docker](https://github.com/docker/docker-install)
Allows to run software in containers. Required for using the provided Docker Image.

<a name="prepro"></a>
## 4.2 Preprocessing 
These tools are used before transcriptome reconstruction, to basecall, evaluate and refine the raw sequencing data.
<a name="dorado"></a>
### 4.2.1 [dorado](https://github.com/nanoporetech/dorado)
Used to convert raw ONT .pod5 files into .ubam basecalled reads.
<a name="pycoQC"></a>
### 4.2.2 [pycoQC](https://github.com/a-slide/pycoQC)
Generates metrics and interactive plots for ONT sequencing data.
<a name="NanoPlot"></a>
### 4.2.3 [NanoPlot](https://github.com/wdecoster/NanoPlot)
Generates metrics and plots for ONT seqeuncing data.
<a name="NanoComp"></a>
### 4.2.4 [NanoComp](https://github.com/wdecoster/nanocomp)
Compares multiple runs of long read sequencing data and alignments based on QC metrics and plots.

<a name="trans-recon"></a>
## 4.3 Transcriptome Reconstruction 
these tools are used to reconstruct the transcriptome from the ONT dRNA sequencing reads using a reference genome file.
<a name="minimap2"></a>
### 4.3.1 [minimap2](https://github.com/lh3/minimap2)
Aligns the raw sequencing reads to the reference genome in spliced mode.

<a name="samtools"></a>
### 4.3.2 [samtools](https://github.com/samtools/samtools)
Sorts and indexes the alignment files (.bam/.sam) generated by minimap2.

<a name="stringtie2"></a>
### 4.3.3 [stringtie2](https://github.com/skovaka/stringtie2)
Reconstruct the transcriptome from the minimap2 mappings and outputs a GTF file containing transcript structure information.

<a name="gffread"></a>
### 4.3.4 [gffread](https://github.com/gpertea/gffread)
Extracts transcript sequences from the genome assembly using the stringtie GTF output and generates a fasta file.

<a name="trans-anno"></a>
## 4.4 Transcriptome Annotation
These tools annotate the reconstructed transcripts with functional information.

<a name="eggNOG"></a>
### 4.3.1 [eggNOG](https://github.com/eggnogdb/eggnog-mapper)
Annotates transcripts using orthology-based functional information.

<a name="blast"></a>
### 4.3.2 [Blast](https://github.com/enormandeau/ncbi_blast_tutorial)
Used to annotate transcripts with UniProt database, identify sequence similarity and reciprocal best hits between transcripts.
